clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -1871,4 +1873,100 @@ void handleFailure(Throwable throwable) {
+                    KafkaFutureImpl<Long> future = futures.get(topicError.getKey());
+                    future.completeExceptionally(topicError.getValue().exception());
+                }
+
+                // grouping topic partitions per leader
+                Map<Node, Map<TopicPartition, Long>> leaders = new HashMap<>();
+                for (Map.Entry<TopicPartition, RecordsToDelete> entry: partitionsAndOffsets.entrySet()) {
+                    Node node = response.cluster().leaderFor(entry.getKey());
+                    if (node != null) {
+                        if (!leaders.containsKey(node))
+                            leaders.put(node, new HashMap<TopicPartition, Long>());
+                        leaders.get(node).put(entry.getKey(), entry.getValue().offset());
+                    } else {
+                        KafkaFutureImpl<Long> future = futures.get(entry.getKey());
+                        future.completeExceptionally(Errors.LEADER_NOT_AVAILABLE.exception());
+                    }
+                }
+
+                for (final Map.Entry<Node, Map<TopicPartition, Long>> entry: leaders.entrySet()) {
+
+                    final long nowDelete = time.milliseconds();
+
+                    final int brokerId = entry.getKey().id();
+
+                    runnable.call(new Call("deleteRecords", deadline,
+                            new ConstantNodeIdProvider(brokerId)) {
+
+                        @Override
+                        AbstractRequest.Builder createRequest(int timeoutMs) {
+                            return new DeleteRecordsRequest.Builder(timeoutMs, entry.getValue());
+                        }
+
+                        @Override
+                        void handleResponse(AbstractResponse abstractResponse) {
+                            DeleteRecordsResponse response = (DeleteRecordsResponse) abstractResponse;
+                            for (Map.Entry<TopicPartition, DeleteRecordsResponse.PartitionResponse> result: response.responses().entrySet()) {
+
+                                KafkaFutureImpl<Long> future = futures.get(result.getKey());
+                                if (result.getValue().error == Errors.NONE) {
+                                    future.complete(result.getValue().lowWatermark);
+                                } else {
+                                    future.completeExceptionally(result.getValue().error.exception());
+                                }
+                            }
+                        }
+
+                        @Override
+                        void handleFailure(Throwable throwable) {
+                            completeAllExceptionally(futures.values(), throwable);
+                        }
+                    }, nowDelete);
+                }
+            }
+
+            @Override
+            void handleFailure(Throwable throwable) {
+                completeAllExceptionally(futures.values(), throwable);
```

Comment:
Ditto here. I think we should consider getting rid of the metadata request and exposing any exceptions from this request to the user's expected delete record response, instead we just rely on the whatever the current metadata (up-to-date or not) and if there is no leader known we set the future exception immediately.

Reply:
Please, let's not.  The other functions in AdminClient do not rely on metadata caching-- they use the latest metadata that is available.  Deleting records shouldn't be a common operation.  If it is, we can have a metadata cache with a configurable expiration time.  I think it's also really bad to set an exception based on possibly stale information.  You give the user no way out if the cache is stale (besides creating an entirely new admin client object, I suppose).

Reply:
As per our offline discussion, we'll leave it like this for now